wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/internal/DagChecker.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema.internal
+
+/**
+ * Checks whether a graph is a directed acyclic graph using [Tarjan's algorithm][tarjan].
```

Comment:
ideas for interview question

Reply:
I definitely needed a lot of time to get this right, including reading the code sample from Wikipedia repeatedly. I would fail an interview that asks me to produce this!